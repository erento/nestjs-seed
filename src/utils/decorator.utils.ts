/* tslint:disable:variable-name */
import {Header} from '@nestjs/common';
type DecoratorFunction = () => MethodDecorator;

export const PrivateCache: DecoratorFunction = (): MethodDecorator =>
    <T>(target: Object, propertyKey: string | symbol, descriptor: TypedPropertyDescriptor<T>): void => {
        Header('Cache-Control', 'private')(target, propertyKey, descriptor);
        Header('Surrogate-Control', 'private')(target, propertyKey, descriptor);
    };
