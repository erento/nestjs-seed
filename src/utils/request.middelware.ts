import {ExpressMiddleware, Middleware, NestMiddleware} from '@nestjs/common';
import {ErentoLogger} from '../common/logger';

@Middleware()
export class RequestMiddelware implements NestMiddleware {
    constructor (private logger: ErentoLogger) {}

    public async resolve (): Promise<ExpressMiddleware> {
        return async (req: Request, _res: any, next: any): Promise<any> => {
            this.logger.log(
                req.url,
                req.mode,
                JSON.stringify(req['query']),
                JSON.stringify(req['params']),
            );
            next();
        };
    }
}
