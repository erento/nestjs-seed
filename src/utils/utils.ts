// tslint:disable typedef

import {UtilsError} from './utils-error';

interface ES7Array extends Array<any> {
    includes (value: any): boolean;
}

export const createBuffer = (body: string, encoding: BufferEncoding = 'base64'): Buffer => {
    try {
        return Buffer.from(body, encoding);
    } catch {
        throw new UtilsError('Cannot create a buffer.');
    }
};

export const deepCopyObj = <T>(sourceObj: T): T | undefined => {
    try {
        return JSON.parse(JSON.stringify(sourceObj));
    } catch {
        return undefined;
    }
};

// Array.prototype.includes is supported since node@6. It is safe to use it.
export const arrayIntersection = (arr1: any[], arr2: any[]) => arr1.filter(x => (<ES7Array> arr2).includes(x));

export const arrayDifference = (arr1: any[], arr2: any[]) => arr1.filter(x => !(<ES7Array> arr2).includes(x))
    .concat(arr2.filter(x => !(<ES7Array> arr1).includes(x)));
