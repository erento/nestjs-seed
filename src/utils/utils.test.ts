import {createBuffer, deepCopyObj} from './utils';

describe('utils', () => {
    describe('buffer', () => {
        test('create', () => {
            const base64: string = 'YWhveQ==';
            const buffer: Buffer | undefined = createBuffer(base64);

            expect(buffer).toBeInstanceOf(Buffer);
            expect(buffer.toString()).toBe('ahoy');
        });

        test('create with wrong data', () => {
            expect(createBuffer('not a base 64')).toBeInstanceOf(Buffer);
            expect(() => createBuffer(<any> null)).toThrowErrorMatchingSnapshot();
            expect(() => createBuffer(<any> undefined)).toThrowErrorMatchingSnapshot();
        });
    });

    describe('object', () => {
        test('copy', () => {
            const objA: any = {a: {a: 1}, b: {b: 1}, c: 1};
            const copy: any = deepCopyObj(objA);
            expect(copy).toMatchSnapshot();
            objA.a.a = 'changed value';
            expect(copy).toMatchSnapshot();
        });

        test('copy not an object', () => {
            expect(deepCopyObj(undefined)).toBeUndefined();
            expect(deepCopyObj(null)).toBeNull();
            expect(deepCopyObj(NaN)).toBeNull();
            expect(deepCopyObj('')).toBe('');
            expect(deepCopyObj('some string')).toBe('some string');
            expect(deepCopyObj(1)).toBe(1);
        });
    });
});
