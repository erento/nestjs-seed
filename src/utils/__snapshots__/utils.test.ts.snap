// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils buffer create with wrong data 1`] = `"Cannot create a buffer."`;

exports[`utils buffer create with wrong data 2`] = `"Cannot create a buffer."`;

exports[`utils object copy 1`] = `
Object {
  "a": Object {
    "a": 1,
  },
  "b": Object {
    "b": 1,
  },
  "c": 1,
}
`;

exports[`utils object copy 2`] = `
Object {
  "a": Object {
    "a": 1,
  },
  "b": Object {
    "b": 1,
  },
  "c": 1,
}
`;
